rfcs/0000-windows-containers.md

```diff
@@ -0,0 +1,119 @@
+
+## Outcomes:
+
+### kpack should run on clusters that have windows workers
+
+kpack service pods and build pods for linux images will need to know not to run on the windows workers to work in a mixed-os cluster.
+
+The kpack release configuration should specify to run pods explicitly on linux os.
+The kpack controller should schedule linux builds on linux workers.
+
+**Actions to Take**
+
+Use the following node selector for all service pods and build pods for linux images:
+
+```
+nodeSelector:
+    kubernetes.io/os: linux
+```
+
+### kpack should support creating windows builders
+
+This is currently supported in the [pack cli](https://github.com/buildpacks/pack/issues/469)
+
+**Actions to Take**
+
+- kpack-controller will use a new environment variable `WINDOWS_LIFECYCLE_IMAGE` to reference the windows lifecycle image
+- When a Builder or ClusterBuilder is created using a windows stack and windows buildpackages, kpack-controller must create a windows builder using the windows lifecycle image.
+    - Image os is determined by the image [os property](https://github.com/opencontainers/image-spec/blob/master/config.md#:~:text=os%20string).
+
+Notes:
+
+- Builders with mixed os images must error gracefully
+- Builders should be labeled with `os.kpack.io: windows` for os disambiguation independent of image metadata
+
+### kpack should support running builds on windows workers
+
+**Actions to Take**
+
+- kpack-controller will accept new environment variables for windows images needed to run builds. These will have the same functionality as on linux and will be used for running builds on windows workers.
+    - `WINDOWS_BUILD_INIT_IMAGE` - a windows build init image
+    - `WINDOWS_COMPLETION_IMAGE` - a windows completion image
+    - `WINDOWS_REBASE_IMAGE` - a windows rebase image
+    - These will be optional to remove the dependency on windows images
+- When an Image is created using a windows Builer or ClusterBuilder labeled with `os.kpack.io: windows`, it will create a build with new properties
+    - The Build will use `WINDOWS_BUILD_INIT_IMAGE` and `WINDOWS_COMPLETION_IMAGE` for the respective containers.
+    - The Build will be labeled with `os.kpack.io: windows` for os disambiguation
+ - When a Build is created with the label `os.kpack.io: windows`, it will use the node selector
+
+ ```
+ nodeSelector:
+     kubernetes.io/os: windows
+ ```
+
+## Complexity:
+
+High
+
+## Prior Art:
+
+[pack cli](https://github.com/buildpacks/pack) has support for windows containers
+
+## Risks:
+
+**Windows images can only run and be created on workers with the same OS version**
+
+Windows images can only run on windows nodes with the same OS. ex: Windows Server LTSC 2019 can only run images that are ltsc2019 or equivalent version [docs](https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility?tabs=windows-server-2019%2Cwindows-10-20H2#:~:text=Supports%20process%20isolation)
```

Comment:
Is there a specific windows version we are targeting and will we need windows build-init/rebase/completion images for each targeted windows version?

Reply:
Given the following:
- `Windows Server 2019 (LTSC) aka version 1809 is the only Windows operating system supported currently on all k8s api versions`
- `Once they support Windows containers with Hyper-V isolation in Kubernetes, the limitation and compatibility rules will change.`

I think we should only support one windows version (LTSC 2019) and kpack should only be able to support one version at a time

Reply:
I'd recommend starting with support for Windows Hosts and Images at version LTSC2019. It's currently the only version you can use as a [k8s node](https://kubernetes.io/docs/setup/production-environment/windows/intro-windows-in-kubernetes/#windows-os-version-support) that *also* has long-term support from Microsoft ([supported until 2024](https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/base-image-lifecycle#:~:text=17763,01/09/2024), whereas other Semi-Annual Channel (SAC) host/image versions only get 18 months of support).

Handling non-matching host/image OS versions gets fuzzier after that... LTSC2019 nodes cannot *currently* create k8s containers from newer images [but may eventually](https://kubernetes.io/docs/setup/production-environment/windows/intro-windows-in-kubernetes/#:~:text=Once%20we%20support%20Windows%20containers%20with%20Hyper). Microsoft does have strong binary ABI compatibility which should ensure binaries created in an LTSC2019 container would run on a 2004 host/image.

With that said, it's probably worth validating that anyone would be running non-LTSC2019 versions in production clusters, I suspect there may not be a strong use case for supporting the SAC versions.